### ISSUE TITLE
Cobench user library

### ISSUE THREAD

**Issue by art-w** [ocurrent/current-bench#318]
The expected "json output" on `make bench` can be a blocker for new users, since it's not documented enough. It also limits the schema changes we can do to add new functionalities.

It would be nice to have a friendly library like `Alcotest` to ease the mental cost of creating benchmarks.

1. [ ] The first version can be really simple: Just a function to output a metric in the expected format!
   ```ocaml
   val metric : name:string
             -> ?units:[`seconds | ...]
             -> ?trend:[`lower_is_better | ...]
             -> ?descr:string
             -> ...
             -> float
             -> unit
   ```

2. [ ] Document how to setup a dune rule `@bench` and a Makefile bench rule?

3. [ ] Would be nice to have a `test_case` interface, where the elapsed time and other metrics are recorded for the user. The target here is the "common case" benchmarks for libraries that just record the raw performance of their operations (ie have a default setup that provides a lot value, with timings, GC info, etc.)

4. [ ] Add an alternative way to push the metrics with the new API

We jokingly talked about this as `Alcobench`, but unless it provides a really colored output, it's not fair to steal their name! Perhaps `Cobench` for Continuous Benchmarking would fit better? Any other ideas?

**Comment by shubhi-arora** [comment]
Hi @art-w , I am an outreachy applicant, i want to work on this issue, but need guidance on how to start with it, can you please help me out?

**Comment by art-w** [comment]
Hello @shubhi-arora, thanks for taking a look at this!

We have a very rough draft of the code for this issue in the [`cobench` folder](https://github.com/ocurrent/current-bench/tree/main/cobench) of this repository, with a lot of things missing! A good start would be to add a new function `val print : t -> unit` that outputs its argument in json format (with the existing [`to_json`](https://github.com/ocurrent/current-bench/blob/main/cobench/cobench.mli#L23) + [`Yojson.Safe.to_string`](https://ocaml-community.github.io/yojson/yojson/Yojson/Safe/index.html#val-to_string) + [`print_endline`](https://v2.ocaml.org/api/Stdlib.html#VALprint_endline))

If you are getting started with OCaml, you'll need to first install [`opam`](https://opam.ocaml.org/) in order to then install [`dune`](https://v3.ocaml.org/p/dune/3.0.3) (to build and run the code) and the json library [`yojson`](https://v3.ocaml.org/p/yojson/1.7.0).  See also [these tutorials for getting started](https://v3.ocaml.org/docs/up-and-running) :)

Feel free to ask as many questions as you need!

**Comment by gs0510** [comment]
I'm going to take number 1 on this issue :)

**Comment by ElectreAAS** [comment]
@art-w do you think this is still necessary after #422 and Cb-check?
I'm asking because cobench is the only reason we have a dependency on `mtime`, which proves to be a little annoying

**Comment by art-w** [comment]
In its current form, I don't think anyone uses cobench -- but I do think it would be easier for projects to use a library than to learn the json format? (cb-check answers the problem of validation, not so much of producing?)
